include/rmm/mr/device/fixed_multisize_memory_resource.hpp

```diff
@@ -0,0 +1,211 @@
+   * @param max_size The maximum fixed block size to allocate
+   * @param initial_blocks_per_size The number of blocks to preallocate from the upstream memory
+   *        resource
+   */
+  explicit fixed_multisize_memory_resource(
+      Upstream* upstream_resource,
+      std::size_t min_size = default_min_size,
+      std::size_t max_size = default_max_size,
+      std::size_t initial_blocks_per_size = 128)
+      : upstream_mr_{upstream_resource},
+        min_size_{min_size},
+        max_size_{max_size} {
+    RMM_EXPECTS(rmm::detail::is_pow2(min_size), "Minimum size must be a power of two");
+    RMM_EXPECTS(rmm::detail::is_pow2(max_size), "Maximum size must be a power of two");
+    
+    // allocate initial blocks and insert into free list
+    for (std::size_t i = min_size; i <= max_size; i *= 2) {
+      fixed_size_mr_.emplace_back(new fixed_size_memory_resource<Upstream>(
+          upstream_resource, i, initial_blocks_per_size));
+    }
+  }
+
+  /**
+   * @brief Destroy the fixed_multisize_memory_resource and free all memory allocated from the
+   *        upstream resource.
+   */
+  virtual ~fixed_multisize_memory_resource() {}
+    
+  /**
+   * @brief Query whether the resource supports use of non-null streams for
+   * allocation/deallocation.
+   *
+   * @returns true
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   * 
+   * @return bool true if the resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override {return false; }
+
+  /**
+   * @brief Get the upstream memory_resource object.
+   *
+   * @return UpstreamResource* the upstream memory resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_mr_; }
+
+  /**
+   * @brief Get the minimum block size that this memory_resource can allocate.
+   * 
+   * @return std::size_t The minimum block size this memory_resource can allocate.
+   */
+  std::size_t get_min_size() const noexcept { return min_size_; }
+
+  /**
+   * @brief Get the maximum block size that this memory_resource can allocate.
+   * 
+   * @return std::size_t The maximum block size this memory_resource can allocate.
+   */
+  std::size_t get_max_size() const noexcept { return max_size_; }
+
+ private:
+
+  /**
+   * @brief Get the memory resource for the requested size
+   *
+   * Chooses a memory_resource that allocates the smallest blocks larger than `bytes`.
+   *
+   * @throws rmm::bad_alloc if the size is larger than the maximum block size.
+   *
+   * @param bytes Requested allocation size in bytes
+   * @return rmm::mr::device_memory_resource& memory_resource that can allocate the requested size.
+   */
+  device_memory_resource* get_resource(std::size_t bytes) {
+    assert(bytes <= max_size_);
+    if (bytes <= max_size_) {
+      auto iter = std::find_if(
+          fixed_size_mr_.begin(), fixed_size_mr_.end(),
```

Comment:
If `fixed_size_mr_` was a `std::set` ordered by the block size of each MR then this operation could be a binary search.

Reply:
It can be a binary search with a vector, too. I switched this to `upper_bound`.

Reply:
Actually, neither works, because I would need to be able to implicitly convert from a `size_t` to a `fixed_size_memory_resource`...  but there's a better way -- what I really want is a binary search for the index...

Reply:
Done. (not a std::set, but using upper_bound with transform_iterator to get the index of the resource with a binary search)